src/sync.cpp

```diff
@@ -203,8 +220,19 @@ void AssertLockHeldInternal(const char *pszName, const char *pszFile, int nLine,
     abort();
 }
 
-// BU normally CCriticalSection is a typedef, but when lockorder debugging is on we need to delete the critical section
-// from the lockorder map
+void AssertWriteLockHeldInternal(const char *pszName, const char *pszFile, int nLine, CSharedCriticalSection *cs)
+{
+    if (cs->try_lock()) // It would be better to check that this thread has the lock
+    {
+        fprintf(stderr, "Assertion failed: lock %s not held in %s:%i; locks held:\n%s", pszName, pszFile, nLine,
+            LocksHeld().c_str());
+        fflush(stderr);
+        abort();
+    }
```

Comment:
the name of the function is AssertWriteLockHeld.... which implies that you'll throw an assert somewhere but there is no assertion.

Reply:
The function calls abort().  We don't want to call the g++ assert() here because we are controlling whether the function is a no-op, not the NDEBUG flag.   
(I was copying the behavior of the existing AssertLockHeldInternal -- see above)